### PR TITLE
T25169 Allow unsetting OARS filter

### PR DESCRIPTION
**Pull request by pwithnall**
This also includes two trivial patches cherry-picked from upstream to fix a couple of unrelated compiler warnings I saw while building this.

https://phabricator.endlessm.com/T25169

### CODE REVIEW

File: panels/user-accounts/um-app-permissions.c [hunk 1]
```diff
@@ -54,7 +55,7 @@ struct _UmAppPermissions
 
   GCancellable *cancellable; /* (owned) */
   EpcAppFilter *filter; /* (owned) */
-  guint         selected_age;
+  guint         selected_age;  /* `(guint32) -1` to disable OARS */
```
Comment (GeorgesStavracas) [inline panels/user-accounts/um-app-permissions.c]: Nitpick: remove one space before `/*`

File: panels/user-accounts/um-app-permissions.c [hunk 2]
```diff
@@ -250,10 +251,19 @@ update_categories_from_language (UmAppPermissions *self)
   /* Fill in the age menu */
   g_menu_remove_all (self->age_menu);
 
+    {
+      g_autofree gchar *action = g_strdup_printf ("permissions.set-age(uint32 %u)", (guint32) -1);
```
Comment (GeorgesStavracas) [inline panels/user-accounts/um-app-permissions.c]: Nitpick: move the variable declaration to the top of the function.

File: panels/user-accounts/um-app-permissions.c [hunk 3]
```diff
@@ -250,10 +251,19 @@ update_categories_from_language (UmAppPermissions *self)
...
+      g_menu_append (self->age_menu, _("Disabled"), action);
+    }
+
   for (i = 0; entries[i] != NULL; i++)
     {
       g_autofree gchar *action = g_strdup_printf ("permissions.set-age(uint32 %u)", ages[i]);
 
+      /* Prevent the unlikely case that one of the real ages is the same as our
+       * special ‘disabled’ value. */
+      g_assert (ages[i] != (guint32) -1);
```
Comment (GeorgesStavracas) [inline panels/user-accounts/um-app-permissions.c]: I'd rather move all these `(guint32) -1` to a `#define AGE_DISABLED ((guint32) -1)` (I'm sure there would be a better name for that though!). To make this and the lines below a bit more semantic.

File: panels/user-accounts/um-app-permissions.c [hunk 4]
```diff
@@ -297,6 +308,8 @@ update_oars_level (UmAppPermissions *self)
       guint age;
 
       oars_value = epc_app_filter_get_oars_value (self->filter, oars_categories[i]);
+      if (oars_value != EPC_APP_FILTER_OARS_VALUE_UNKNOWN)
+        all_categories_unset = FALSE;
```
Comment (GeorgesStavracas) [inline panels/user-accounts/um-app-permissions.c]: I'd rather do:

```
      all_categories_unsed &= oars_value != EPC_APP_FILTER_OARS_VALUE_UNKNOWN;
```

File: panels/user-accounts/um-app-permissions.c [hunk 5]
```diff
@@ -286,6 +296,7 @@ update_oars_level (UmAppPermissions *self)
   const gchar *rating_age_category;
   guint maximum_age;
   gsize i;
+  gboolean all_categories_unset = TRUE;
```
Comment (GeorgesStavracas) [inline panels/user-accounts/um-app-permissions.c]: Nitpick: please move `all_categories_unset = TRUE;` to below `maximum_age = 0;` at line 303

File: panels/user-accounts/um-app-permissions.c (hunk 1 above)
Comment (pwithnall) [reply panels/user-accounts/um-app-permissions.c]: I was using the Python style, but whatever.

File: panels/user-accounts/um-app-permissions.c (hunk 4 above)
Comment (pwithnall) [reply panels/user-accounts/um-app-permissions.c]: Actually `all_categories_unsed &= oars_value == EPC_APP_FILTER_OARS_VALUE_UNKNOWN;`

**Comment by pwithnall** [comment]
Pushed a fixup with the nitpick fixes; ready for re-review.

Do **not** merge until the fixup is squashed.

File: panels/user-accounts/um-app-permissions.c [hunk 6]
```diff
@@ -22,12 +22,16 @@
...
 #include <gio/gdesktopappinfo.h>
+#include <glib/gi18n.h>
 #include <strings.h>
 
 #include "gs-content-rating.h"
 
 #include "um-app-permissions.h"
 
+/* The value which we store as an age to indicate that OARS filtering is disabled. */
+static const guint32 oars_disabled_age = (guint32) -1;
```
Comment (GeorgesStavracas) [inline panels/user-accounts/um-app-permissions.c]: Why not a `#define`?

Comment (pwithnall) [reply panels/user-accounts/um-app-permissions.c]: A `static const` allows the compiler to do its own type checking and optimisation. A `#define` is just a string substitution which the compiler doesn’t get to know anything about, which is almost always worse for compilation.

**Comment by pwithnall** [comment]
Squashed. Merging once Jenkins passes.

**Pull request merged**